README.md
# phlib/sms-length

[![Code Checks](https://img.shields.io/github/workflow/status/phlib/sms-length/CodeChecks?logo=github)](https://github.com/phlib/sms-length/actions/workflows/code-checks.yml)
[![Codecov](https://img.shields.io/codecov/c/github/phlib/sms-length.svg?logo=codecov)](https://codecov.io/gh/phlib/sms-length)
[![Latest Stable Version](https://img.shields.io/packagist/v/phlib/sms-length.svg?logo=packagist)](https://packagist.org/packages/phlib/sms-length)
[![Total Downloads](https://img.shields.io/packagist/dt/phlib/sms-length.svg?logo=packagist)](https://packagist.org/packages/phlib/sms-length)
![Licence](https://img.shields.io/github/license/phlib/sms-length.svg)

Calculate SMS GSM 03.38 message encoding and length, and number of concatenated SMS required

## Installation

```sh
composer require phlib/sms-length
```

## Usage

Simple string which fits within GSM 03.38 7-bit alphabet:

```php
$smsLength = new \Phlib\SmsLength\SmsLength('simple message');
$smsLength->validate(); // Throw exceptions for any validation failures
$smsLength->getSize(); // 14
$smsLength->getEncoding(); // '7-bit'
$smsLength->getMessageCount(); // 1
$smsLength->getUpperBreakpoint(); // 160
```

Message which contains characters forcing switch to using GSM 03.38 UCS-2:

```php
$smsLength = new \Phlib\SmsLength\SmsLength('message with • char requiring UCS-2');
$smsLength->validate(); // Throw exceptions for any validation failures
$smsLength->getSize(); // 35
$smsLength->getEncoding(); // 'ucs-2'
$smsLength->getMessageCount(); // 1
$smsLength->getUpperBreakpoint(); // 70
```

## Background

In the course of adding an SMS module to our Commercial software, we have found
need for the SMS length properties which we've included in this package.

Our third-party SMS gateway provider's API will of course error if the message
length exceeds the maximum allowed, and handles the correct concatenation and
encoding for the given UTF-8 string.

However we want to be able to inform our users of message size and limits while
they're building their SMS campaign (in advance of a live send). Plus we also
need to show how many concatenated SMS they will use per contact, and therefore
indicate pricing.

The unit tests and inspections made by this package are based on
[GSM 03.38 / 3GPP 23.038](https://en.wikipedia.org/wiki/GSM_03.38) for encoding
and [GSM 03.40 / 3GPP 23.040](https://en.wikipedia.org/wiki/GSM_03.40) for
[concatenated SMS](https://en.wikipedia.org/wiki/Concatenated_SMS). We have also
referred to documentation provided by third-party SMS gateway providers such as
[MessageBird](https://support.messagebird.com/hc/en-us/articles/208739745-How-long-can-a-text-message-be-)
and [Messente](https://messente.com/documentation/tools/sms-length-calculator).

## License

This package is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
## [2.0.1] - 2022-06-26
### Added
- Use constants for encoding types. HT: SeifertDaniel

## [2.0.0] - 2021-07-18
### Added
- Add specific support for PHP v8
- Type declarations have been added to all method parameters and return types.
### Removed
- **BC break**: Removed support for PHP versions <= v7.2 as they are no longer
  [actively supported](https://php.net/supported-versions.php) by the PHP project.

## [1.1.0] - 2017-04-18
### Added
- New `validate()` method to check the content string matches the specification.
### Changed
- Updated construct docblock to make it clear that the SMS message content
should be provided as a UTF-8 string. In most cases this should not require any
implementation changes as UTF-8 is the
[default charset for PHP](http://php.net/manual/en/ini.core.php#ini.default-charset).
- The exception thrown for message content which exceeds the maximum allowed
number of concatenated SMS (255) has been moved from the inspection on construct
to a dedicated `validate()` method. This allows the implementation to still read
the other stats from the constructed object. (The constructor will still throw
an exception if the message string contained characters that cannot be parsed).

## [1.0.0] - 2017-03-28
Initial Release


tests/SmsLengthTest.php
<?php

declare(strict_types=1);

namespace Phlib\SmsLength;

use Phlib\SmsLength\Exception\InvalidArgumentException;
use PHPUnit\Framework\TestCase;

class SmsLengthTest extends TestCase
{
    /**
     * @see https://en.wikipedia.org/wiki/GSM_03.38
     * @var string[] Printable characters in GSM 03.38 7-bit default alphabet
     *               0x1B deliberately excluded as it's used to escape to extension table
     */
    private const GSM0338_BASIC =
        '@'  . 'Δ' . ' ' . '0' . '¡' . 'P' . '¿' . 'p' .
        '£'  . '_' . '!' . '1' . 'A' . 'Q' . 'a' . 'q' .
        '$'  . 'Φ' . '"' . '2' . 'B' . 'R' . 'b' . 'r' .
        '¥'  . 'Γ' . '#' . '3' . 'C' . 'S' . 'c' . 's' .
        'è'  . 'Λ' . '¤' . '4' . 'D' . 'T' . 'd' . 't' .
        'é'  . 'Ω' . '%' . '5' . 'E' . 'U' . 'e' . 'u' .
        'ù'  . 'Π' . '&' . '6' . 'F' . 'V' . 'f' . 'v' .
        'ì'  . 'Ψ' . "'" . '7' . 'G' . 'W' . 'g' . 'w' .
        'ò'  . 'Σ' . '(' . '8' . 'H' . 'X' . 'h' . 'x' .
        'Ç'  . 'Θ' . ')' . '9' . 'I' . 'Y' . 'i' . 'y' .
        "\n" . 'Ξ' . '*' . ':' . 'J' . 'Z' . 'j' . 'z' .
        'Ø'        . '+' . ';' . 'K' . 'Ä' . 'k' . 'ä' .
        'ø'  . 'Æ' . ',' . '<' . 'L' . 'Ö' . 'l' . 'ö' .
        "\r" . 'æ' . '-' . '=' . 'M' . 'Ñ' . 'm' . 'ñ' .
        'Å'  . 'ß' . '.' . '>' . 'N' . 'Ü' . 'n' . 'ü' .
        'å'  . 'É' . '/' . '?' . 'O' . '§' . 'o' . 'à'
    ;

    /**
     * @see https://en.wikipedia.org/wiki/GSM_03.38
     * @var string[] Printable characters in GSM 03.38 7-bit extension table
     */
    private const GSM0338_EXTENDED = '|^€{}[~]\\';

    /**
     * @dataProvider providerSize
     */
    public function testSize(
        string $content,
        string $encoding,
        int $characters,
        int $messageCount,
        int $upperBreak
    ): void {
        $size = new SmsLength($content);

        static::assertSame($encoding, $size->getEncoding());
        static::assertSame($characters, $size->getSize());
        static::assertSame($messageCount, $size->getMessageCount());
        static::assertSame($upperBreak, $size->getUpperBreakpoint());

        static::assertTrue($size->validate());
    }

    public function providerSize(): array
    {
        return [
            'gsm-basic' => [self::GSM0338_BASIC, SmsLength::ENCODING_7BIT, 127, 1, 160],
            'gsm-extended' => [self::GSM0338_EXTENDED, SmsLength::ENCODING_7BIT, 18, 1, 160],
            'simple-basic' => ['simple msg', SmsLength::ENCODING_7BIT, 10, 1, 160],
            'simple-extended' => ['simple msg plus € extended char', SmsLength::ENCODING_7BIT, 32, 1, 160],

            // http://www.fileformat.info/info/unicode/char/2022/index.htm
            'utf16-1' => ['simple msg plus • 1 UTF-16 code unit char', SmsLength::ENCODING_UCS2, 41, 1, 70],

            // http://www.fileformat.info/info/unicode/char/1f4f1/index.htm
            'utf16-2' => ["simple msg plus \xf0\x9f\x93\xb1 2 UTF-16 code unit char", SmsLength::ENCODING_UCS2, 42, 1, 70],

            // long 7-bit
            'long-gsm-simple' => [str_repeat('simple msg', 50), SmsLength::ENCODING_7BIT, 500, 4, 612],
            'long-gsm-exact' => [str_repeat('exact max', 153), SmsLength::ENCODING_7BIT, 1377, 9, 1377],

            // long 7-bit extended
            'long-gsm-ex-1' => [str_repeat(self::GSM0338_EXTENDED, 40), SmsLength::ENCODING_7BIT, 720, 5, 765],
            'long-gsm-ex-2' => [str_repeat(self::GSM0338_EXTENDED, 76), SmsLength::ENCODING_7BIT, 1368, 9, 1377],

            // long UCS-2
            'long-ucs-1' => [str_repeat('simple msg plus •', 20), SmsLength::ENCODING_UCS2, 340, 6, 402],
            'long-ucs-2' => [str_repeat("simple msg plus \xf0\x9f\x93\xb1", 20), SmsLength::ENCODING_UCS2, 360, 6, 402],
            'long-ucs-exact' => [str_repeat('exact•max', 67), SmsLength::ENCODING_UCS2, 603, 9, 603],

            // empty
            'empty' => ['', SmsLength::ENCODING_7BIT, 0, 1, 160],
        ];
    }

    /**
     * @dataProvider providerTooLarge
     * @medium Expect tests to take >1 but <10
     */
    public function testTooLarge(
        string $content,
        string $encoding,
        int $characters,
        int $messageCount,
        int $upperBreak
    ): void {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Message count cannot exceed 255');

        $size = new SmsLength($content);

        static::assertSame($encoding, $size->getEncoding());
        static::assertSame($characters, $size->getSize());
        static::assertSame($messageCount, $size->getMessageCount());
        static::assertSame($upperBreak, $size->getUpperBreakpoint());

        // Trigger exception
        $size->validate();
    }

    public function providerTooLarge(): array
    {
        // 7-bit max is 39015 (255 * 153)
        // ucs-2 max is 17085 (255 * 67)
        return [
            // long 7-bit, 10 * 3902 = 39020
            'basic' => [str_repeat('simple msg', 3902), SmsLength::ENCODING_7BIT, 39020, 256, 39168],

            // long 7-bit extended, 18 * 2168 = 39024
            'extended' => [str_repeat(self::GSM0338_EXTENDED, 2168), SmsLength::ENCODING_7BIT, 39024, 256, 39168],

            // long UCS-2 single, 17 * 1006 = 17102
            // long UCS-2 double, 18 * 950 = 17100
            'ucs-1' => [str_repeat('simple msg plus •', 1006), SmsLength::ENCODING_UCS2, 17102, 256, 17152],
            'ucs-2' => [str_repeat("simple msg plus \xf0\x9f\x93\xb1", 950), SmsLength::ENCODING_UCS2, 17100, 256, 17152],
        ];
    }
}


ecs.php
<?php

declare(strict_types=1);

use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Symplify\EasyCodingStandard\ValueObject\Option;
use Symplify\EasyCodingStandard\ValueObject\Set\SetList;

return static function (ContainerConfigurator $containerConfigurator): void {
    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [
        __DIR__ . '/src',
        __DIR__ . '/tests',
    ]);

    $services = $containerConfigurator->services();

    $containerConfigurator->import(SetList::COMMON);
    // Remove sniff, from common/control-structures
    $services->remove(\PhpCsFixer\Fixer\ClassNotation\OrderedClassElementsFixer::class);
    // Remove sniff, from common/spaces
    $services->remove(\PhpCsFixer\Fixer\Operator\NotOperatorWithSuccessorSpaceFixer::class);
    $services->remove(\PhpCsFixer\Fixer\CastNotation\CastSpacesFixer::class);

    $containerConfigurator->import(SetList::PSR_12);
};


src/SmsLength.php
<?php

declare(strict_types=1);

namespace Phlib\SmsLength;

use Phlib\SmsLength\Exception\InvalidArgumentException;

/**
 * @package    Phlib\SmsLength
 * @license    LGPL-3.0
 */
class SmsLength
{
    /**
     * @var string name for 7-bit encoding
     */
    public const ENCODING_7BIT = '7-bit';
    
    /**
     * @var string name for UCS-2 encoding
     */
    public const ENCODING_UCS2 = 'ucs-2';
    
    /**
     * @var int Maximum characters in SMS with 7-bit encoding (3GPP TS 23.038 / GSM 03.38)
     */
    public const MAXIMUM_CHARACTERS_7BIT_SINGLE = 160;

    /**
     * @var int Maximum characters in SMS with 7-bit encoding with UDH (3GPP TS 23.040)
     */
    public const MAXIMUM_CHARACTERS_7BIT_CONCATENATED = 153;

    /**
     * @var int Maximum characters in SMS with UCS-2 encoding (3GPP TS 23.038 / GSM 03.38)
     */
    public const MAXIMUM_CHARACTERS_UCS2_SINGLE = 70;

    /**
     * @var int Maximum characters in SMS with UCS-2 encoding with UDH (3GPP TS 23.040)
     */
    public const MAXIMUM_CHARACTERS_UCS2_CONCATENATED = 67;

    /**
     * @var int Message cannot exceed size of 255 concatenated SMS (3GPP TS 23.040)
     */
    public const MAXIMUM_CONCATENATED_SMS = 255;

    /**
     * @see https://en.wikipedia.org/wiki/GSM_03.38
     * @var string[] Printable characters in GSM 03.38 7-bit default alphabet
     *               0x1B deliberately excluded as it's used to escape to extension table
     */
    private const GSM0338_BASIC = [
        '@' , 'Δ', ' ', '0', '¡', 'P', '¿', 'p',
        '£' , '_', '!', '1', 'A', 'Q', 'a', 'q',
        '$' , 'Φ', '"', '2', 'B', 'R', 'b', 'r',
        '¥' , 'Γ', '#', '3', 'C', 'S', 'c', 's',
        'è' , 'Λ', '¤', '4', 'D', 'T', 'd', 't',
        'é' , 'Ω', '%', '5', 'E', 'U', 'e', 'u',
        'ù' , 'Π', '&', '6', 'F', 'V', 'f', 'v',
        'ì' , 'Ψ', "'", '7', 'G', 'W', 'g', 'w',
        'ò' , 'Σ', '(', '8', 'H', 'X', 'h', 'x',
        'Ç' , 'Θ', ')', '9', 'I', 'Y', 'i', 'y',
        "\n", 'Ξ', '*', ':', 'J', 'Z', 'j', 'z',
        'Ø'      , '+', ';', 'K', 'Ä', 'k', 'ä',
        'ø' , 'Æ', ',', '<', 'L', 'Ö', 'l', 'ö',
        "\r", 'æ', '-', '=', 'M', 'Ñ', 'm', 'ñ',
        'Å' , 'ß', '.', '>', 'N', 'Ü', 'n', 'ü',
        'å' , 'É', '/', '?', 'O', '§', 'o', 'à',
    ];

    /**
     * @see https://en.wikipedia.org/wiki/GSM_03.38
     * @var string[] Printable characters in GSM 03.38 7-bit extension table
     */
    private const GSM0338_EXTENDED = ['|', '^', '€', '{', '}', '[', '~', ']', '\\'];

    /**
     * @var string
     */
    private $encoding;

    /**
     * @var int
     */
    private $size;

    /**
     * @var int
     */
    private $messageCount;

    /**
     * @param string $messageContent SMS message content (UTF-8)
     */
    public function __construct(string $messageContent)
    {
        $this->inspect($messageContent);
    }

    /**
     * Get name of GSM 03.38 encoding that would be required for the given content
     *
     * @return string '7-bit' or 'ucs-2'
     */
    public function getEncoding(): string
    {
        return $this->encoding;
    }

    /**
     * Get size of message as characters used in the determined encoding
     */
    public function getSize(): int
    {
        return $this->size;
    }

    /**
     * Get number of messages that would be used to send the given content size
     */
    public function getMessageCount(): int
    {
        return $this->messageCount;
    }

    /**
     * Get upper breakpoint for the current message count
     */
    public function getUpperBreakpoint(): int
    {
        $single = self::MAXIMUM_CHARACTERS_7BIT_SINGLE;
        $concat = self::MAXIMUM_CHARACTERS_7BIT_CONCATENATED;
        if ($this->encoding === self::ENCODING_UCS2) {
            $single = self::MAXIMUM_CHARACTERS_UCS2_SINGLE;
            $concat = self::MAXIMUM_CHARACTERS_UCS2_CONCATENATED;
        }

        if ($this->size <= $single) {
            return $single;
        }

        return $this->messageCount * $concat;
    }

    /**
     * Check the message content is valid for an SMS
     */
    public function validate(): bool
    {
        if ($this->messageCount > self::MAXIMUM_CONCATENATED_SMS) {
            throw new InvalidArgumentException('Message count cannot exceed ' . self::MAXIMUM_CONCATENATED_SMS);
        }

        return true;
    }

    /**
     * Parse content to discover size characteristics
     */
    private function inspect(string $messageContent): void
    {
        // If it's not UTF-8, then it's broken
        if (!mb_check_encoding($messageContent, 'UTF-8')) {
            throw new InvalidArgumentException('Content encoding could not be verified ' . $messageContent);
        }

        // Start counting characters in basic and extended
        // Extended chars count for two
        // Any character outside the 7-bit alphabet switches the entire encoding to UCS-2
        $this->encoding = self::ENCODING_7BIT;
        $this->size = 0;
        $mbLength = mb_strlen($messageContent, 'UTF-8');
        for ($i = 0; $i < $mbLength; $i++) {
            $char = mb_substr($messageContent, $i, 1, 'UTF-8');
            if (in_array($char, self::GSM0338_BASIC, true)) {
                $this->size++;
            } elseif (in_array($char, self::GSM0338_EXTENDED, true)) {
                $this->size += 2;
            } else {
                $this->encoding = self::ENCODING_UCS2;
                break;
            }
        }

        if ($this->encoding === self::ENCODING_UCS2) {
            // For UCS-2 need to iterate the characters again
            // Those with two UTF-16 code points consume two characters in the SMS
            $this->size = 0;
            for ($i = 0; $i < $mbLength; $i++) {
                $char = mb_substr($messageContent, $i, 1, 'UTF-8');
                $utf16Hex = bin2hex(mb_convert_encoding($char, 'UTF-16', 'UTF-8'));
                $this->size += strlen($utf16Hex) / 4;
            }
        }

        // Message Count: Each SMS is slightly shorter if concatenation is required
        $singleSize = self::MAXIMUM_CHARACTERS_7BIT_SINGLE;
        $concatSize = self::MAXIMUM_CHARACTERS_7BIT_CONCATENATED;
        if ($this->encoding === self::ENCODING_UCS2) {
            $singleSize = self::MAXIMUM_CHARACTERS_UCS2_SINGLE;
            $concatSize = self::MAXIMUM_CHARACTERS_UCS2_CONCATENATED;
        }

        $this->messageCount = 1;
        if ($this->size > $singleSize) {
            $this->messageCount = (int)ceil($this->size / $concatSize);
        }
    }
}
